api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java

```diff
@@ -0,0 +1,204 @@
+   *
+   * @param singleHeader whether to use single or multiple headers.
+   */
+  public B3Propagator(boolean singleHeader) {
+    this.singleHeader = singleHeader;
+  }
+
+  @Override
+  public List<String> fields() {
+    return FIELDS;
+  }
+
+  @Override
+  public <C> void inject(SpanContext spanContext, C carrier, Setter<C> setter) {
+    checkNotNull(spanContext, "spanContext");
+    checkNotNull(setter, "setter");
+    checkNotNull(carrier, "carrier");
+
+    String sampled = spanContext.getTraceFlags().isSampled() ? TRUE_INT : FALSE_INT;
+
+    if (singleHeader) {
+      setter.set(
+          carrier,
+          COMBINED_HEADER,
+          spanContext.getTraceId().toLowerBase16()
+              + COMBINED_HEADER_DELIMITER
+              + spanContext.getSpanId().toLowerBase16()
+              + COMBINED_HEADER_DELIMITER
+              + sampled);
+    } else {
+      setter.set(carrier, TRACE_ID_HEADER, spanContext.getTraceId().toLowerBase16());
+      setter.set(carrier, SPAN_ID_HEADER, spanContext.getSpanId().toLowerBase16());
+      setter.set(carrier, SAMPLED_HEADER, sampled);
+    }
+  }
+
+  @Override
+  public <C /*>>> extends @NonNull Object*/> SpanContext extract(C carrier, Getter<C> getter) {
+    checkNotNull(carrier, "carrier");
+    checkNotNull(getter, "getter");
+
+    if (singleHeader) {
+      return getSpanContextFromSingleHeader(carrier, getter);
+    } else {
+      return getSpanContextFromMultipleHeaders(carrier, getter);
+    }
+  }
+
+  @SuppressWarnings("StringSplitter")
+  private static <C /*>>> extends @NonNull Object*/> SpanContext getSpanContextFromSingleHeader(
+      C carrier, Getter<C> getter) {
+    String value = getter.get(carrier, COMBINED_HEADER);
+    if (StringUtils.isNullOrEmpty(value)) {
+      logger.info(
+          "Missing or empty combined header: "
+              + COMBINED_HEADER
+              + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    // must have between 2 and 4 hyphen delimieted parts:
+    //   traceId-spanId-sampled-parentSpanId (last two are optional)
+    // NOTE: we do not use parentSpanId
+    String[] parts = value.split(COMBINED_HEADER_DELIMITER);
+    if (parts.length < 2 || parts.length > 4) {
+      logger.info(
+          "Invalid combined header '" + COMBINED_HEADER + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String traceId = parts[0];
+    if (!isTraceIdValid(traceId)) {
+      logger.info(
+          "Invalid TraceId in B3 header: " + COMBINED_HEADER + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String spanId = parts[1];
+    if (!isSpanIdValid(spanId)) {
+      logger.info(
+          "Invalid SpanId in B3 header: " + COMBINED_HEADER + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String sampled = parts.length == 3 ? parts[2] : null;
```

Comment:
Based on my understanding sampled is present if parts.length is 3 or 4 :)

Reply:
You're correct. We need to check the sampled string if we have three or more items in the array.